modules/sfp_abstractapi.py

```diff
@@ -0,0 +1,181 @@
+            'references': [
+                "https://app.abstractapi.com/api/ip-geolocation/documentation"
+            ],
+            'apiKeyInstructions': [
+                "Visit https://app.abstractapi.com/users/signup",
+                "Register a free account",
+                "Browse to https://app.abstractapi.com/api/ip-geolocation/tester",
+                "Click on 'Try it out'",
+                "Your API Key will be listed under 'This is your private API key, specific to this API.'",
+            ],
+            'favIcon': "https://global-uploads.webflow.com/5ebbd0a566a3996636e55959/5ec2ba27ede983917dbff22f_favicon.png",
+            'logo': "https://global-uploads.webflow.com/5ebbd0a566a3996636e55959/5ec2b974e578f93e553425eb_logo-dark.svg",
+            'description': "The IP Geolocation allows you to look up the location, timezone, country details, and more of an IPv4 or IPv6 address.",
+        }
+    }
+
+    # Default options
+    opts = {
+        "api_key": ""
+    }
+
+    # Option descriptions
+    optdescs = {
+        "api_key": "AbstractAPI Geolocation API Key"
+    }
+
+    results = None
+
+    def setup(self, sfc, userOpts=dict()):
+        self.sf = sfc
+        self.results = self.tempStorage()
+
+        for opt in list(userOpts.keys()):
+            self.opts[opt] = userOpts[opt]
+
+    # What events is this module interested in for input
+    def watchedEvents(self):
+        return [
+            "IP_ADDRESS",
+            "IPV6_ADDRESS"
+        ]
+
+    # What events this module produces
+    # This is to support the end user in selecting modules based on events
+    # produced.
+    def producedEvents(self):
+        return [
+            "GEOINFO",
+            "PHYISCAL_COORDINATES",
+            "RAW_RIR_DATA"
+        ]
+
+    def query(self, qry):
+        queryString = f"https://ipgeolocation.abstractapi.com/v1/?api_key={self.opts['api_key']}&ip_address={qry}"
+
+        res = self.sf.fetchUrl(queryString,
+                               timeout=self.opts['_fetchtimeout'],
+                               useragent=self.opts['_useragent'])
+        time.sleep(1.2)
+
+        return self.parseAPIResponse(res)
+
+    def parseAPIResponse(self, res):
+        if res['code'] == '204':
+            self.sf.debug("No geolocation information for target")
+            return None
+
+        if res['code'] == "400":
+            self.sf.error("Bad Request")
+            self.errorState = True
+            return None
+
+        if res['code'] == "401":
+            self.sf.error("Unauthorized. Invalid abstractapi API key")
+            self.errorState = True
+            return None
+
+        if res['code'] == '422':
+            self.sf.error("Usage quota reached. Insufficient API Credit")
+            self.errorState = True
+            return None
+
+        if res['code'] == '429':
+            self.sf.error("Too many requests")
+            self.errorState = True
```

Comment:
Not sure this should cause `errorState` to be `True` since it indicates one throttled request, not that all future requests will fail (that's the only condition under which `errorState` should be `True`.

Reply:
Updated